nengo_ocl/simulator.py

```diff
@@ -850,22 +872,37 @@ def plan_ElementwiseInc(self, ops):
         return [plan_elementwise_inc(self.queue, A, X, Y)]
 
     def plan_SparseDotInc(self, ops):
-        plans = []
+        ''' This currently gives one plan per sparse operation instead of combining them all '''
+        scipy_As = [self.sparse_data[self.sparse_sidx[op.A]] for op in ops]
+        return self._plan_csr_multiply(ops, scipy_As)
+
+    def plan_BsrDotInc(self, ops):
+        ''' Technically this will be implemented as CSR multiplication,
+            Still, the underlying weights remains referenced, but not the sparsity structure
+        '''
+        scipy_As = []
         for op in ops:
-            assert op.A.sparse
-            A = self.sparse_data[self.sparse_sidx[op.A]]
-            tt = [None]
-            xx = [op.X]
-            yy = [op.Y]
-            fn_name = "%s.fn" % (op,)
-
-            def fn(x, A=A):
-                return A.dot(x)
-
-            plans.append(self._plan_fn_in_python(fn, tt, xx, yy, fn_name))
-
-        return plans
+            bsr_values_raveled = self.all_data[self.sidx[op.A]]
+            bsr_values = np.reshape(bsr_values_raveled, self.bsr_reshapes[op])
+            bsr_A = npext.scipy_sparse.bsr_matrix((bsr_values, op.indices, op.indptr))
+            scipy_As.append(bsr_A.tocsr())
+        return self._plan_csr_multiply(ops, scipy_As)
+
+    def _plan_csr_multiply(self, ops, scipy_As, inc=False):
+        ''' scipy_As is a list of CSR matrices that have been extracted from the ops '''
+        if len(ops) > 1:
+            return [self._plan_csr_multiply([op], [scipy_A])[0] for op, scipy_A in zip(ops, scipy_As)]
+        if any(A.dtype not in [np.int32, np.float32] for A in scipy_As):
+            raise TypeError('Only 32-bit sparse matrices supported with OCL, and casting at this point would break reference to original.\n \
+                             Go to where the matrix is created and make dtype=np.float32')
```

Comment:
This is causing problems with the Nengo tests, because in the Nengo frontend we typically handle everything as `float64`. For standard (dense) data in NengoOCL, we explicitly cast to float32 when we build, so I think we should do the same here (i.e. if `A` is not `float32`, cast it to that). Also, I wouldn't bother supporting the `int32` case (it's actually not currently supported either, because in `plan_sparsedot_inc`, you require `X`, `Y`, and `A` to all be the same type, and `X` and `Y` will always be `float32`).